File: Python/015_3Sum.py
"""
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.
"""

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        for i in range(len(nums) - 2):
            # Handle of repetitions of the first element in the triplet
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            j, k = i + 1, len(nums) - 1
            while j < k and j < len(nums):
                if nums[i] + nums[j] + nums[k] == 0:
                    res.append([nums[i], nums[j], nums[k]])
                    # After finding a triplet, deduplicate the second and third element
                    while j < k and nums[j] == nums[j + 1]:
                        j += 1
                    while j < k and nums[k] == nums[k - 1]:
                        k -= 1    
                    k -= 1
                    j += 1
                elif nums[i] + nums[j] + nums[k] > 0:
                    k -= 1
                else:
                    j += 1   
        return res             

File: Java/015_3Sum.java
/**
 * 
 * Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
 * Notice that the solution set must not contain duplicate triplets.
 */
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
       List<List<Integer>> res = new ArrayList<>();
       Arrays.sort(nums);
       for (int i = 0; i < nums.length; ++i){
           int left = i + 1, right = nums.length - 1;
           if (i > 0 && nums[i] == nums[i - 1]){
               continue;
           }
           while (left < right){
               if (nums[i] + nums[left] + nums[right] == 0){
                   res.add(Arrays.asList(nums[i], nums[left], nums[right]));
                   while (right > left && nums[right] == nums[right - 1]){
                       right--;
                   }
                   while (right > left && nums[left] == nums[left + 1]){
                       left++;
                   }
                   right--;
                   left++;
               }else if (nums[i] + nums[left] + nums[right] > 0){
                   right--;
               }else {
                   left++;
               }
           }
       }
       return res;         
    }
}